sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/BoundAttribute.scala

```diff
@@ -60,17 +60,28 @@ case class BoundReference(ordinal: Int, dataType: DataType, nullable: Boolean)
 
   override def genCode(ctx: CodegenContext, ev: ExprCode): String = {
     val javaType = ctx.javaType(dataType)
-    val value = ctx.getValue(ctx.INPUT_ROW, dataType, ordinal.toString)
+    val value = if (!ctx.isColumnarType(ctx.INPUT_ROW)) {
+      ctx.getValue(ctx.INPUT_ROW, dataType, ordinal.toString)
+    } else {
+      ctx.getValue(ctx.INPUT_ROW, dataType, ctx.INPUT_COLORDINAL)
```

Comment:
One possible idea is to prepare ```getValue()``` function that is can be overridden as follows.
In my opinion, the following code is not easy to read. What do you think?

In ```BoundExpression```
```scala
def getValue(...): String = { ctx.getValue(ctx.INPUT_ROW, dataType, ordinal.toString) }

override genCode(...): String = {
  ...
  val value = getValue(...)
 ...
}
```

In ```DataSourceScan```, we pass ```BoundExpression``` that has own ```getValue()``` as follows:

```scala
...
val exprs = output.zipWithIndex.map(x => new BoundReference(x._2, x._1.dataType, true) {
  def getValue(...): String = {
    val value = if (!ctx.isColumnarType(ctx.INPUT_ROW)) {
      ctx.getValue(ctx.INPUT_ROW, dataType, ordinal.toString)
    } else {
      ctx.getValue(ctx.INPUT_ROW, dataType, ctx.INPUT_COLORDINAL)
    }
  }
})
...
```